File: 2024/day1/part2.cpp

#include <iostream>
#include <vector>
#include <bits/stdc++.h>
#include <unordered_map>


using namespace std;

FILE *inputFile;
int num;
int x = 0;
int ans = 0;
vector<int> list1;
vector<int> list2;
std::unordered_map<int, int> freq;

int main(void) {
    inputFile = fopen("input.txt", "r");
    
    // read input
    while (fscanf(inputFile, "%d", &num) == 1) {
        // alternate between pushing to lists
        if (x % 2 == 0) {
            list1.push_back(num);
        } else {
            list2.push_back(num);
        }
        x++;
    }

    fclose(inputFile);

    // get frequency of all list2 numbers
    for (int num : list2) {
        if (freq.find(num) != freq.end()) {
            freq[num] += 1;
        } else {
            freq[num] = 1;
        }
    }

    // iterate over list1 and multiply it by freq
    for (int num : list1) {
        ans += num * freq[num];
    }

    cout << ans << '\n';
}

File: 2024/day1/part2.py
# read input file and seperate into lists of lines
lines = open('input.txt', 'r').readlines()

dict = {}
list1 = []

ans = 0


for line in lines:
    # seperate based on spaces between elements
    parts = line.split()
    # append to list1
    list1.append(int(parts[0]))
    # get frequency of each dict
    if int(parts[1]) in dict:
        dict[int(parts[1])] += 1
    else:
        dict[int(parts[1])] = 1
    

for item in list1:
    # check if its in dict otherwise 0
    if item in dict:
        ans += item * dict[item]

print(ans)